core/pkg/server/stream.go

```diff
@@ -342,83 +330,94 @@ func (s *Stream) Start() {
 	}()
 
 	// handle dispatching between components
-	s.wg.Add(1)
+	s.wg.Add(2)
 	go func() {
-		wg := sync.WaitGroup{}
 		for _, ch := range []chan *service.Result{s.handler.outChan, s.sender.outChan} {
-			wg.Add(1)
 			go func(ch chan *service.Result) {
 				for result := range ch {
 					s.dispatcher.handleRespond(result)
 				}
-				wg.Done()
+				s.wg.Done()
 			}(ch)
 		}
-		wg.Wait()
-		close(s.outChan)
-		s.wg.Done()
 	}()
 	s.logger.Debug("starting stream", "id", s.settings.GetRunID())
 }
 
+// loopSendInputsToHandler forwards records from inChan and loopbackChan
+// to the handler channel until done.
+func (s *Stream) loopSendInputsToHandler(handlerChan chan<- *service.Record) {
+	for {
+		select {
+		// Select from inChan or loopBackChan while there's buffered data.
+		case x := <-s.inChan:
+			handlerChan <- x
+		case x := <-s.loopBackChan:
+			handlerChan <- x
+
+		// Otherwise, race between inChan / loopBackChan / done.
+		default:
+			select {
+			case x := <-s.inChan:
+				handlerChan <- x
+			case x := <-s.loopBackChan:
+				handlerChan <- x
+			case <-s.done:
+				return
```

Comment:
i think ideally we want to guarantee that we processed all the data before closing the stream?

Reply:
the outer select consumes all buffered data before considering the `done` channel. We only close `done` after Sender cancels the `ctx`, so only after the defer state machine has completed---this means there should not be any more requests.